js/src/components/contact-information/phone-number-card/useCountdown.js

```diff
@@ -0,0 +1,76 @@
+/**
+ * External dependencies
+ */
+import { useState, useEffect, useCallback, useRef } from '@wordpress/element';
+
+/**
+ * @typedef {Object} Countdown
+ * @property {number} second The current remaining seconds of countdown.
+ * @property {number} callCount The calling count of `startCountdown` function.
+ * @property {(countdownSecond: number) => void} startCountdown Start the countdown with the given starting seconds.
+ */
+
+/**
+ * Get the corresponding countdown states and its start function by given `handle`.
+ * The states would be updated every 1000 ms interval till counting down to 0.
+ *
+ * @param {string} handle The countdown handle.
+ * @return {Countdown} The corresponding countdown states and its start function.
+ */
+export default function useCountdown( handle ) {
+	const timerRef = useRef( {} );
+	const [ second, setSecond ] = useState( 0 );
+
+	timerRef.current.usingHandle = handle;
+	timerRef.current[ handle ] = timerRef.current[ handle ] || { callCount: 0 };
+
+	const startCountdown = useCallback(
+		( countdownSecond ) => {
+			setSecond( countdownSecond );
+
+			const timer = timerRef.current[ handle ];
+			const timeUp = new Date().getTime() + 1000 * countdownSecond;
+
+			if ( timer.id ) {
+				clearInterval( timer.id );
+			}
+
+			timer.updateSecond = () => {
+				let remainingSecond = ( timeUp - new Date().getTime() ) / 1000;
+				remainingSecond = Math.max( Math.round( remainingSecond ), 0 );
```

Comment:
❔ The time calculation feels a bit like acrobatic (new Date(), Math.max, Math.round), wondering maybe we can make it simpler by manipulating integer second instead, i.e. `60`, `59`, `58`... just minus `1` every one second interval until it becomes `0`.

Reply:
There're two reasons to calculate it by time factor:

1. I'm not sure if the cooldown time would be a hard rule limited by our or Google's API. If it's limited by API, then it'd be proper to count by time diff.
2. When the process of event loop is blocked, the timer would be blocked as well. Counting by time diff can still result in a more accurate countdown. For example that the timer is blocked by `alert()`:
   ![Kapture 2021-09-07 at 14 39 50](https://user-images.githubusercontent.com/17420811/132296564-bc179a60-721f-4835-ac90-15b2573fb12d.gif)

In most cases, counting by time diff as default preference would not bring surprise when using timers.

Reply:
Thanks for the explanation, I gave some more thoughts about it and yeah it makes sense to use time diff 🙂  👍